File: UCF HSPT Documents/2014/Solutions/sponge.c

#include <stdio.h>

void main() {
   int times, t;

   // Open the input file
   FILE* fin = fopen("sponge.in", "r");
   
   // Read in the number of sponges to process and iterate over them
   fscanf(fin, "%d", &times); 
   for(t = 1; t <= times; t++) {
      int n, s, j;

      // Read in the sponge-level and side-length
      fscanf(fin, "%d %d", &n, &s); 
      
      // Level-0 sponge volume
      double volume = s*s*s; 
      
      // We make the observation that a sponge of level (i+1) 
      // has 20/27s the volume of a sponge of level i.
      // This is because each sub-cube in the level (i+1) sponge
      // is divided into 27 smaller cubes, of which 7 are discarded.
      
      // So, we need merely reduce the level 0 sponge n times to find 
      // the final volume. 
      double ratio = 20.0/27.0; 
      for (j = 0; j < n; j++)
         volume *= ratio;
      
      // Output our answer, rounded to 3 decimal places
      printf("Sponge #%d: %.3f\n", t,volume);
   }

   // Close file and return
   fclose(fin);
   return 0;
}

File: UCF HSPT Project/src/y2014/sponge.java

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class sponge {
   public static void main(String[] args) throws FileNotFoundException {
      // Think of each sponge as 27 smaller sponges. When we make the fractal
      // we cut away 7 of them and leave 20. When we make a level two fractal
      // we do the same thing for each cube. Each cube has 20/27ths the
      // volume. Each time we go a level deeper the size gets multiplied by
      // this ratio. Note that "20d" and "27d" are written with a d so they
      // get treated as doubles.
      final double ratio = 20d / 27d;

      // Open file and read in number of cases
      Scanner in = new Scanner(new File("sponge.in"));
      int cases = in.nextInt();
      int caseNum = 0;

      // Loop over the cases
      while (caseNum++ < cases) {
         // Get input case (level of sponge and the original level-0 size)
         int levels = in.nextInt();
         int originalSize = in.nextInt();

         // The original volume is the original length cubed. We multiply
         // that by ratio once for each level. This can be expressed as one
         // multiplication by ratio to the power of the number of levels.
         double newSize = Math.pow(ratio, levels)
               * Math.pow(originalSize, 3);

         // Output the answer
         System.out.printf("Sponge #%d: %.3f%n", caseNum, newSize);
      }

      // Close input file
      in.close();
   }
}
